src/algebra/monoid_algebra/basic.lean

```diff
@@ -1503,6 +1536,20 @@ finset.induction_on s rfl $ λ a s has ih, by rw [prod_insert has, ih,
 
 end
 
+lemma map_domain_algebra_map {A H F : Type*} [comm_semiring k] [semiring A]
+  [algebra k A] [add_monoid G] [add_monoid H] [add_monoid_hom_class F G H] (f : F) (r : k) :
+  map_domain f (algebra_map k (add_monoid_algebra A G) r) =
+    algebra_map k (add_monoid_algebra A H) r :=
+by simp only [function.comp_app, map_domain_single, add_monoid_algebra.coe_algebra_map, map_zero]
+
+/--  An additive homomorphism `f : G →+ H` induces a `k`-algebra homomorphism
+`map_domain_alg_hom k A f : add_monoid_algebra A G →ₐ[k] add_monoid_algebra A H`. -/
+@[simps] def map_domain_alg_hom (k A : Type*) [comm_semiring k] [semiring A] [algebra k A]
+  [add_monoid G] {H F : Type*} [add_monoid H] [add_monoid_hom_class F G H] (f : F) :
+  add_monoid_algebra A G →ₐ[k] add_monoid_algebra A H :=
```

Comment:
At risk of scope creeping forever... Are you happy to add the `non_unital_alg_hom` version too? It would make sense to merge master to get #13470 if so.

Reply:
I am happy to do it, since I am not in a rush to need this PR, though I may need some assistance with the proofs, since I know nothing of the API surrounding `non_unital_alg_hom`s.

Reply:
I updated the doc-strings, merged master and pushed.

Reply:
Ok, I will give it a go, but I will wait for the oleans to be available.

Reply:
`leanproject get-cache --fallback=download-first` should get you appropriate oleans immediately, since this PR touches only one file

Reply:
I get:
```
$ leanproject get-cache --fallback=download-first
Usage: leanproject get-cache [OPTIONS]
Try 'leanproject get-cache -h' for help.

Error: no such option: --fallback
```

Reply:
In any case, your comment made me realize that I could manually pull the oleans of master and proceed from there.

Reply:
You get that error because you're on a (very!) old version of `leanproject`

Reply:
It seems that `monoid_algebra A G` asks for `A` to be a `semiring`.  Where should I use a `non_unital_non_assoc`-something?

Reply:
You should replace `monoid G` with `has_mul G` I think

Reply:
how do I update `leanproject`?  I am currently on `leanproject, version 1.0.0`.

Reply:
That's probably a question for Zulip. The latest version is 1.1.1.

Reply:
I am trying with
```lean
lemma map_domain_algebra_map (k A : Type*) {G H F : Type*} [comm_semiring k] [semiring A]
  [algebra k A] [has_mul G] [has_mul H] [mul_hom_class F G H] (f : F) (r : k) :
  map_domain f (non_unital_alg_hom k (monoid_algebra A G) r) =
    algebra_map k (monoid_algebra A H) r :=
```
but I cannot get it to typecheck.  I tried various inputs and typeclass assumptions, but nothing worked so far-

Reply:
That doesn't typecheck because it's nonsense, you should be copying `map_domain_alg_hom` but replacing `→ₐ[k]` with `→ₙₐ[k]`. You've copied `map_domain_algebra_map ` instead and ended up in a mess.

Reply:
(I suspect you're being confused by `algebra_map`, which is a function, vs `linear_map` and `alg_hom`, which are the morphism types)

Reply:
It should be something like:
```lean
/-- If `f : G → H` is a homomorphism between two magmas, then `finsupp.map_domain f` is an
non-unital algebra homomorphism between their magma algebras. -/
@[simps]
def map_domain_non_unital_alg_hom (k A : Type*) [comm_semiring k] [semiring A] [algebra k A]
  {H F : Type*} [has_mul H] [has_mul F G H] (f : F) :
  monoid_algebra A G →ₙₐ[k] monoid_algebra A H :=
{ map_mul' := λ x y, map_domain_mul f x y,
  map_smul' := λ r x, map_domain_smul f r x,
  ..(finsupp.map_domain.add_monoid_hom f : monoid_algebra k G →+ monoid_algebra k H) }
  ```

Reply:
Ok, I think that I got it to work!

You are right that I find the names very confusing.  I know that this is my fault for not paying too much attention to them, but navigating this PR is quite hard for me.

Reply:
Anyway, I hope that people will not start removing zeros from rings before this PR gets merged!  :upside_down_face:

Reply:
I just pulled that PR off the queue, so you have a tiny bit more time!